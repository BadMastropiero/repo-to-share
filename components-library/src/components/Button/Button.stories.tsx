import {ComponentMeta, ComponentStory} from '@storybook/react';

import Button from './Button';
import {ArrowLeft, ArrowRight} from '../Icon';

export default {
  title: 'Components/Button',
  component: Button,
} as ComponentMeta<typeof Button>;

type StoryProps = {
  text: string;
  primary: boolean;
  disabled: boolean;
  showAfter: boolean;
  showPrev: boolean;
  presetSize?: 'large' | 'normal' | 'small';
  presetType?: 'text' | 'default';
  onClick?: () => void;
  rounded?: boolean;
};

type StoryAllProps = {
  text: string;
  primary: boolean;
  disabled: boolean;
  rounded?: boolean;
  onClick?: () => void;
};

export const Base: ComponentStory<React.FC<StoryProps>> = args => {
  const {primary, text, showAfter, showPrev, disabled, presetSize, onClick, rounded, presetType} =
    args;

  return (
    <Button
      primary={primary}
      disabled={disabled}
      onClick={onClick}
      text={text}
      afterIcon={showAfter ? ArrowRight : undefined}
      beforeIcon={showPrev ? ArrowLeft : undefined}
      presetSize={presetSize}
      rounded={rounded}
      presetType={presetType}
    />
  );
};

export const Link: ComponentStory<React.FC<StoryProps>> = args => {
  const {primary, text, showAfter, showPrev, disabled, presetSize, onClick, rounded, presetType} =
    args;

  return (
    <Button
      href="http://example.com"
      primary={primary}
      disabled={disabled}
      onClick={onClick}
      text={text}
      afterIcon={showAfter ? ArrowRight : undefined}
      beforeIcon={showPrev ? ArrowLeft : undefined}
      presetSize={presetSize}
      rounded={rounded}
      presetType={presetType}
    />
  );
};

export const All: ComponentStory<React.FC<StoryAllProps>> = args => {
  const {primary, text, disabled, onClick, rounded} = args;

  return (
    <div style={{display: 'flex', gap: '10px', alignItems: 'center'}}>
      <Button
        primary={primary}
        disabled={disabled}
        onClick={onClick}
        text={text}
        afterIcon={ArrowRight}
        presetSize="large"
        rounded={rounded}
      />

      <Button
        primary={primary}
        disabled={disabled}
        onClick={onClick}
        text={text}
        afterIcon={ArrowRight}
        presetSize="normal"
        rounded={rounded}
      />

      <Button
        primary={primary}
        disabled={disabled}
        onClick={onClick}
        text={text}
        afterIcon={ArrowRight}
        presetSize="small"
        rounded={rounded}
      />
    </div>
  );
};

Base.args = {
  primary: true,
  disabled: false,
  text: 'Button',
  showAfter: true,
  showPrev: false,
  presetSize: 'large',
  onClick: () => {},
  rounded: true,
};

Link.args = {
  primary: true,
  disabled: false,
  text: 'Button',
  showAfter: true,
  showPrev: false,
  presetSize: 'large',
  onClick: () => {},
  rounded: true,
};

All.args = {
  primary: true,
  disabled: false,
  text: 'Button',
  onClick: () => {},
  rounded: true,
};

Base.parameters = {
  design: {
    type: 'figma',
    url: 'https://www.figma.com/file/epHXucNHjke2rES69iOsaS/TetraImpacts?node-id=2%3A1157',
  },
};

All.parameters = {
  design: {
    type: 'figma',
    url: 'https://www.figma.com/file/epHXucNHjke2rES69iOsaS/TetraImpacts?node-id=2%3A1157',
  },
};
