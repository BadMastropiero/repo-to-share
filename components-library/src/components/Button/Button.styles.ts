import styled, {css} from 'styled-components';
import {DynamicComponent} from '../DynamicComponent';

export const StyledIconContainer = styled.div<{primary: boolean; padding?: string}>(
  ({
    theme: {
      colors: {primary100, primaryDark},
    },
    primary: isPrimary,
    padding,
  }) => css`
    transition: all 0.3s;
    display: flex;
    flex-direction: row;
    justify-content: center;
    align-items: center;
    background: ${isPrimary ? primaryDark : primary100};
    padding: ${padding};
  `
);

export const StyledButton = styled(DynamicComponent)<{primary: boolean; rounded: boolean}>(
  ({
    theme: {
      colors: {
        primary,
        primaryDark,
        primary50,
        primary100,
        primary200,
        primary600,
        primary800,
        primaryBg,
      },
      transition,
    },
    primary: isPrimary,
    rounded,
  }) => css`
    transition: ${transition};
    position: relative; // For the ripple effect to work
    display: flex;
    flex-direction: row;
    justify-content: center;
    align-items: center;
    padding: 0px;
    isolation: isolate;
    box-sizing: border-box;
    background-color: ${isPrimary ? primary : primary50};
    margin: 0;
    padding: 0;
    box-shadow: 0px 10px 15px -5px rgba(0, 101, 166, 0.4);
    border-radius: ${rounded ? '10px' : null};
    border: 4px solid ${isPrimary ? primary600 : primaryBg};
    overflow: hidden;
    height: min-content;
    &:disabled {
      background-color: ${isPrimary ? primary : primary50};
      ${StyledIconContainer} {
        background: ${isPrimary ? primaryDark : primary100};
      }
    }
    &:hover:not(:disabled) {
      background-color: ${isPrimary ? primaryDark : primary100};
      ${StyledIconContainer} {
        background: ${isPrimary ? primary800 : primary200};
      }
    }
    &:active:not(:disabled) {
      box-shadow: none;
    }
  `
);

export const StyledTextButton = styled(StyledButton)(
  ({
    theme: {
      colors: {primaryDark, primary100},
    },
    primary: isPrimary,
  }) => css`
    background-color: transparent;
    margin: 0;
    padding: 0;
    box-shadow: none;
    border: none;
    overflow: hidden;
    height: min-content;
    &:disabled {
      background-color: transparent;
    }
    &:hover:not(:disabled) {
      // TODO this line is here to update styles if needs to be differ from original button
      background-color: ${!isPrimary ? primaryDark : primary100};
    }
  `
);

export default StyledButton;
