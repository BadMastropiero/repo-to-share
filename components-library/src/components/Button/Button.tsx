import React, {useContext} from 'react';
import {ThemeContext} from 'styled-components';

// Styles
import {StyledButton, StyledIconContainer, StyledTextButton} from './Button.styles';

// Interfaces
import {ButtonProps} from './Button.interface';
import {IconSizeProp} from '../Icon/Icon.interface';
// Components
import {Typography} from '../Typography';
import {Icon} from '../Icon';
import Ripple from '../Ripple/Ripple';

const mapSizeToTypographyPreset = {
  large: 'buttonLG',
  normal: 'button',
  small: 'buttonSM',
};

const mapSizeToIconPadding = {
  large: '14px',
  normal: '10px',
  small: '6px',
};

const mapSizeToIconSize = {
  large: 'normal',
  normal: 'normalBtn',
  small: 'small',
};

const mapPresetTypeToButtonType = {
  text: StyledTextButton,
  default: StyledButton,
};

const Button: React.FC<ButtonProps> = ({
  primary = false,
  disabled = false,
  afterIcon,
  afterIconComponent,
  beforeIcon,
  beforeIconComponent,
  text,
  textStyle,
  presetSize = 'normal',
  rounded = true,
  className,
  presetType = 'default',
  style,
  href,
  ...rest
}) => {
  const {colors} = useContext(ThemeContext);

  const ButtonToRender = mapPresetTypeToButtonType[presetType] || mapPresetTypeToButtonType.default;

  let textColor = 'secondary';
  if ((primary && presetType !== 'text') || (presetType === 'text' && !primary)) {
    textColor = disabled ? colors.primaryDark : 'secondary';
  } else {
    textColor = disabled ? colors.primary200 : colors.primaryDark;
  }

  let iconColor = null;
  if (primary) {
    iconColor = disabled ? colors.primary : null;
  } else {
    iconColor = disabled ? colors.primaryBg : colors.primary;
  }

  return (
    <ButtonToRender
      tag={href ? 'link' : 'button'}
      href={href}
      className={`ti-button ${className || ''}`}
      primary={primary}
      disabled={disabled}
      rounded={rounded}
      style={{
        textDecoration: href ? 'unset' : undefined,
        cursor: href ? 'pointer' : undefined,
        width: href ? 'fit-content' : undefined,
        ...style,
      }}
      {...rest}
    >
      {beforeIcon && !beforeIconComponent && presetType !== 'text' && (
        <StyledIconContainer
          primary={primary}
          padding={presetSize ? mapSizeToIconPadding[presetSize] : undefined}
        >
          <Icon
            icon={beforeIcon}
            primary={!primary}
            disabled={false}
            customColor={iconColor}
            sizePreset={
              (presetSize
                ? mapSizeToIconSize[presetSize] || 'normalBtn'
                : 'normalBtn') as IconSizeProp
            }
          />
        </StyledIconContainer>
      )}
      {beforeIconComponent}
      {text && (
        <Typography
          variant="p"
          customPreset={presetSize ? mapSizeToTypographyPreset[presetSize] || 'button' : 'button'}
          customColor={textColor}
          style={textStyle}
        >
          {text}
        </Typography>
      )}
      {afterIcon && !afterIconComponent && presetType !== 'text' && (
        <StyledIconContainer
          primary={primary}
          padding={presetSize ? mapSizeToIconPadding[presetSize] : undefined}
        >
          <Icon
            icon={afterIcon}
            primary={!primary}
            disabled={false}
            customColor={iconColor}
            sizePreset={
              (presetSize
                ? mapSizeToIconSize[presetSize] || 'normalBtn'
                : 'normalBtn') as IconSizeProp
            }
          />
        </StyledIconContainer>
      )}
      {afterIconComponent}
      <Ripple color={primary ? colors.primary200 : colors.primaryBg} duration={1000} />
    </ButtonToRender>
  );
};

export default Button;
