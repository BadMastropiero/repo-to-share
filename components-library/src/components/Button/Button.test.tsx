import {render, screen} from '@testing-library/react';
import {composeStories} from '@storybook/testing-react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import * as stories from './Button.stories';

// Every component that is returned maps 1:1 with the stories, but they already contain all decorators from story level, meta level and global level!
const {Base} = composeStories(stories);

describe('Button tests', () => {
  test('should render', () => {
    render(<Base />);
    expect(screen.getByRole('button')).toBeInTheDocument();
  });

  test('should match snapshot', () => {
    const tree = render(<Base />);
    expect(tree).toMatchSnapshot();
  });

  it('should trigger on click', () => {
    const mockCallBack = jest.fn();

    render(<Base onClick={mockCallBack} />);
    const button = screen.getByRole('button');
    userEvent.click(button);
    expect(mockCallBack.mock.calls.length).toEqual(1);
  });
});
