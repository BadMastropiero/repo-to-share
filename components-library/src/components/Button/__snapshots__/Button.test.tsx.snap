// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button tests should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="sc-bcXHqe sc-dkrFOg gkrZeY ti-button "
      >
        <p
          class="sc-bcXHqe sc-eDvSVe fulKPl"
        >
          Button
        </p>
        <div
          class="sc-gswNZR byNAsO"
        >
          <svg
            class="sc-jSUZER rbNgI ticons-arrow_right"
            height="24px"
            role="img"
            viewBox="0 0 24 24"
            width="24px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6.475 21.5C6.175 21.2 6.025 20.8333 6.025 20.4C6.025 19.9667 6.175 19.6 6.475 19.3L13.8 11.975L6.45 4.625C6.15 4.325 6 3.96667 6 3.55C6 3.13333 6.15833 2.76667 6.475 2.45C6.775 2.15 7.14167 2 7.575 2C8.00833 2 8.375 2.15 8.675 2.45L17.275 11.05C17.4083 11.1833 17.5043 11.3293 17.563 11.488C17.621 11.646 17.65 11.8083 17.65 11.975C17.65 12.1417 17.621 12.304 17.563 12.462C17.5043 12.6207 17.4083 12.7667 17.275 12.9L8.65 21.525C8.35 21.825 7.99167 21.975 7.575 21.975C7.15833 21.975 6.79167 21.8167 6.475 21.5Z"
            />
          </svg>
        </div>
        <div
          class="sc-gKPRtg dOMrEj"
          color="#C5CFDF"
        />
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="sc-bcXHqe sc-dkrFOg gkrZeY ti-button "
    >
      <p
        class="sc-bcXHqe sc-eDvSVe fulKPl"
      >
        Button
      </p>
      <div
        class="sc-gswNZR byNAsO"
      >
        <svg
          class="sc-jSUZER rbNgI ticons-arrow_right"
          height="24px"
          role="img"
          viewBox="0 0 24 24"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M6.475 21.5C6.175 21.2 6.025 20.8333 6.025 20.4C6.025 19.9667 6.175 19.6 6.475 19.3L13.8 11.975L6.45 4.625C6.15 4.325 6 3.96667 6 3.55C6 3.13333 6.15833 2.76667 6.475 2.45C6.775 2.15 7.14167 2 7.575 2C8.00833 2 8.375 2.15 8.675 2.45L17.275 11.05C17.4083 11.1833 17.5043 11.3293 17.563 11.488C17.621 11.646 17.65 11.8083 17.65 11.975C17.65 12.1417 17.621 12.304 17.563 12.462C17.5043 12.6207 17.4083 12.7667 17.275 12.9L8.65 21.525C8.35 21.825 7.99167 21.975 7.575 21.975C7.15833 21.975 6.79167 21.8167 6.475 21.5Z"
          />
        </svg>
      </div>
      <div
        class="sc-gKPRtg dOMrEj"
        color="#C5CFDF"
      />
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
